numpy/core/src/multiarray/item_selection.c

```diff
@@ -483,14 +485,20 @@ PyArray_PutMask(PyArrayObject *self, PyObject* values0, PyObject* mask0)
         Py_INCREF(Py_None);
         return Py_None;
     }
+    src = PyArray_DATA(values);
+
     if (PyDataType_REFCHK(PyArray_DESCR(self))) {
-        for (i = 0; i < ni; i++) {
-            tmp = ((npy_bool *)(PyArray_DATA(mask)))[i];
-            if (tmp) {
-                src = PyArray_BYTES(values) + chunk * (i % nv);
-                PyArray_Item_INCREF(src, PyArray_DESCR(self));
-                PyArray_Item_XDECREF(dest+i*chunk, PyArray_DESCR(self));
-                memmove(dest + i * chunk, src, chunk);
+        for (i = 0, j = 0; i < ni; i++, j++) {
+            if (j >= nv) {
+                j = 0;
+            }
+            if (mask_data[i]) {
+                char *src_ptr = src + j*chunk;
```

Comment:
Could just increment `src` and `dest` by `chunk` each time around, resetting `src`along with `j`. That would make a difference for dense masks, although there are a other bits in the loop that might take time.  Not needed for this PR, though.